File: 2026/rtti/cpp/src/type_registry.cpp
#include "rtti/type_registry.h"

#include <algorithm>
#include <cassert>

namespace rtti {

TypeRegistry& TypeRegistry::instance() {
    static TypeRegistry reg;
    return reg;
}

TypeInfo* TypeRegistry::define(std::string_view name, TypeInfo* parent) {
    return define(name, parent, 0);
}

TypeInfo* TypeRegistry::define(std::string_view name, TypeInfo* parent,
                                std::size_t object_size) {
    if (sealed_) {
        throw std::logic_error("TypeRegistry: already sealed");
    }

    std::string key(name);
    if (by_name_.contains(key)) {
        throw std::logic_error("TypeRegistry: duplicate type '" + key + "'");
    }

    auto* info = new TypeInfo();
    info->name = key;
    info->parent = parent;
    info->id = 0;
    info->high = 0;
    by_name_[key] = info;

    if (parent == nullptr) {
        roots_.push_back(info);
    }

    if (object_size > 0) {
        by_size_[object_size] = info;
    }

    return info;
}

void TypeRegistry::seal() {
    if (sealed_) return;

    ordered_.clear();
    int next_id = 1;

    for (auto* root : roots_) {
        dfs_assign(root, next_id);
    }

    int max_id = next_id - 1;
    by_id_.resize(static_cast<std::size_t>(max_id) + 1);
    for (auto* info : ordered_) {
        by_id_[static_cast<std::size_t>(info->id)] = info;
    }

    sealed_ = true;
}

int TypeRegistry::dfs_assign(TypeInfo* node, int& next_id) {
    int id = next_id++;
    int high = id;

    node->id = id;
    node->high = high;

    std::vector<TypeInfo*> children;
    collect_children(node, children);
    std::sort(children.begin(), children.end(),
              [](const TypeInfo* a, const TypeInfo* b) {
                  return a->name < b->name;
              });

    for (auto* child : children) {
        int child_high = dfs_assign(child, next_id);
        high = child_high;
        node->high = high;
    }

    ordered_.push_back(node);
    return high;
}

void TypeRegistry::collect_children(const TypeInfo* parent,
                                     std::vector<TypeInfo*>& out) {
    for (auto& [name, info] : by_name_) {
        if (info->parent == parent) {
            out.push_back(info);
        }
    }
}

const TypeInfo* TypeRegistry::lookup(int type_id) const noexcept {
    if (type_id < 0 || static_cast<std::size_t>(type_id) >= by_id_.size()) {
        return nullptr;
    }
    return by_id_[static_cast<std::size_t>(type_id)];
}

const TypeInfo* TypeRegistry::lookup(std::string_view name) const noexcept {
    std::string key(name);
    auto it = by_name_.find(key);
    return (it != by_name_.end()) ? it->second : nullptr;
}

bool TypeRegistry::is_assignable_from(const TypeInfo* super,
                                       int type_id) const noexcept {
    return super != nullptr && super->is_assignable_from(type_id);
}

} // namespace rtti

File: 2026/rtti/java/src/main/java/rtti/TypeRegistry.java
package rtti;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

public final class TypeRegistry {

    private static final TypeRegistry INSTANCE = new TypeRegistry();

    private final Map<String, TypeInfo> byName = new ConcurrentHashMap<>();
    private final Map<Class<?>, TypeInfo> byClass = new ConcurrentHashMap<>();
    private volatile TypeInfo[] byId = new TypeInfo[0];
    private volatile boolean sealed = false;

    public TypeRegistry() {}

    public static TypeRegistry get() { return INSTANCE; }

    public synchronized TypeInfo define(String name, TypeInfo parent) {
        return define(name, parent, null);
    }

    public synchronized TypeInfo define(String name, TypeInfo parent, Class<?> javaClass) {
        if (sealed) throw new IllegalStateException("Registry already sealed");
        if (byName.containsKey(name)) throw new IllegalArgumentException("Duplicate type: " + name);

        TypeInfo type = new TypeInfo(0, 0, name, parent, javaClass);
        byName.put(name, type);
        if (javaClass != null) {
            byClass.put(javaClass, type);
        }
        return type;
    }

    public synchronized void seal() {
        if (sealed) return;

        List<TypeInfo> roots = byName.values().stream()
                .filter(t -> t.parent() == null)
                .sorted(Comparator.comparing(TypeInfo::name))
                .collect(Collectors.toList());

        List<TypeInfo> ordered = new ArrayList<>();
        AtomicInteger counter = new AtomicInteger(1);

        for (TypeInfo root : roots) {
            dfsAssign(root, ordered, counter);
        }

        int maxId = counter.get() - 1;
        byId = new TypeInfo[maxId + 1];
        for (TypeInfo t : ordered) {
            byId[t.id()] = t;
        }

        sealed = true;
    }

    private int dfsAssign(TypeInfo node, List<TypeInfo> ordered, AtomicInteger counter) {
        int id = counter.getAndIncrement();
        int high = id;

        node.setId(id);
        node.setHigh(high);

        List<TypeInfo> children = byName.values().stream()
                .filter(t -> t.parent() != null && t.parent().name().equals(node.name()))
                .sorted(Comparator.comparing(TypeInfo::name))
                .collect(Collectors.toList());

        for (TypeInfo child : children) {
            int childHigh = dfsAssign(child, ordered, counter);
            high = childHigh;
            node.setHigh(high);
        }

        ordered.add(node);
        return high;
    }

    public TypeInfo lookup(int typeId) {
        TypeInfo[] snapshot = byId;
        if (typeId < 0 || typeId >= snapshot.length) return null;
        return snapshot[typeId];
    }

    public TypeInfo lookup(String name) {
        return byName.get(name);
    }

    public TypeInfo lookup(Class<?> clazz) {
        return byClass.get(clazz);
    }

    public boolean isAssignableFrom(TypeInfo superType, int typeId) {
        return superType != null && superType.isAssignableFrom(typeId);
    }

    public boolean isSealed() { return sealed; }

    public List<TypeInfo> allTypes() {
        return Collections.unmodifiableList(Arrays.asList(byId).subList(1, byId.length));
    }

    @Override
    public String toString() {
        if (!sealed) return "TypeRegistry[unsealed, " + byName.size() + " types]";
        StringBuilder sb = new StringBuilder("TypeRegistry[");
        for (int i = 1; i < byId.length; i++) {
            if (i > 1) sb.append(", ");
            sb.append(byId[i]);
        }
        sb.append("]");
        return sb.toString();
    }
}
